### PR TITLE
Make oracle penalty settings  user-modifiable at the backend

### LINKED ISSUE #5143: Make oracle penalty info user configurable
## Problem
https://github.com/rotki/rotki/pull/5133 introduced a mechanism that penalizes misbehaving oracle(s) in runtime.

However, the approach hardcodes the constants that determine which oracle(s) should be penalized.
This should not be the case, it should be user configurable.

## Specification
Instead of hardcoding it, move it to `DBSettings` so it can be updated/changed from time to time by the user.

### PR DESCRIPTION
**Pull request by programmer-ke**
Approach:
- Added settings `oracle_penalty_threshold_count` and `oracle_penalty_duration` to replace the constants that were being used by the `PenalizablePriceOracleMixin`.
- Made the settings updateable via the REST API.
- Made the settings accessible from `CachedSettings`.

I noticed that the previously `CachedSettings` was getting initialized by default settings, and updated with any settings updated in the logged in session. However this means that the next time the user logs in, cached settings would have default values but not any saved custom settings.

Changed this so that `CachedSettings` is updated with saved settings when the user is unlocked, so that previously customized settings are also cached.

Closes #5143 

## Checklist

- [ ] The PR modified the frontend, and updated the [user guide](https://github.com/rotki/rotki/blob/develop/docs/usage_guide.rst) to reflect the changes.

### CODE REVIEW

File: rotkehlchen/api/v1/schemas.py
Comment (yabirgb) [inline rotkehlchen/api/v1/schemas.py:1275]: for consistency the message should use the same format in the string

```suggestion
            error='The count should be >= 1',
        ),
    )
    oracle_penalty_duration = fields.Integer(
        load_default=None,
        validate=webargs.validate.Range(
            min=1,
            error='The penalty should be >= 1 seconds',
```

File: rotkehlchen/rotkehlchen.py
Comment (yabirgb) [inline rotkehlchen/rotkehlchen.py:315]: instead of opening two cursors move the update of the cached settings to where the settings were queried before

File: rotkehlchen/tests/api/test_settings.py
Comment (yabirgb) [inline rotkehlchen/tests/api/test_settings.py:38]: This is something that is really only used here. Instead you can use the fixture that we have to set settings. An example of how to use it is

```
@pytest.mark.parametrize('db_settings', [
    {'pnl_csv_with_formulas': False},
])
```

and you can modify some of them and test them

File: rotkehlchen/tests/fixtures/rotkehlchen.py
Comment (yabirgb) [inline rotkehlchen/tests/fixtures/rotkehlchen.py:501]: remove this logic

File: rotkehlchen/utils/mixins/penalizable_oracle.py
Comment (yabirgb) [inline rotkehlchen/utils/mixins/penalizable_oracle.py:21]: instead of all the settings can't you just query the key?

```suggestion
        if self.query_failures_count >= CachedSettings().get_entry('oracle_penalty_threshold_count'):
```

File: rotkehlchen/tests/unit/test_inquirer.py
Comment (yabirgb) [inline rotkehlchen/tests/unit/test_inquirer.py:561]: you did it here

File: rotkehlchen/rotkehlchen.py
Comment (yabirgb) [inline rotkehlchen/rotkehlchen.py:314]: You need to update the cachedsettings docstring since the behaviour has changed

File: rotkehlchen/db/settings.py
Comment (yabirgb) [inline rotkehlchen/db/settings.py:415]: don't leave the empty line. Also why not make both a tuple?
```suggestion
    def update_entries(self, settings: ModifiableDBSettings | DBSettings) -> None:
        if isinstance(settings, ModifiableDBSettings):
            setting_fields = settings._fields
        else:
            setting_fields = tuple(field.name for field in fields(settings))

        for attr in setting_fields:
```

File: docs/api.rst
Comment (yabirgb) [inline docs/api.rst:694]:
```suggestion
   :resjson int oracle_penalty_threshold_count: The number of failures after which an oracle is penalized. Default is 5.
   :resjson int oracle_penalty_duration: The duration in seconds for which an oracle is penalized. Default is 1800.
```

File: docs/api.rst
Comment (yabirgb) [inline docs/api.rst:648]:
```suggestion
   :resjson int oracle_penalty_threshold_count: The number of failures after which an oracle is penalized. Default is 5.
   :resjson int oracle_penalty_duration: The duration in seconds for which an oracle is penalized. Default is 1800.
```

**Review by yabirgb — COMMENTED** [review]
Thank you! SOme tests failed. Please check them

File: rotkehlchen/utils/mixins/penalizable_oracle.py
Comment (programmer-ke) [reply rotkehlchen/utils/mixins/penalizable_oracle.py:21]: Okay, the reason for this was that the type checker wasn't pleased with the use of `get_entry`, because it returns
a union of types yet the one I need is an int.

It would therefore flag the comparison operations as invalid as they don't work with the non-integer types. But I could be missing a way of making it work in case you have a suggestion. Another approach would be to add a property method for each setting I need.

File: rotkehlchen/tests/unit/test_inquirer.py
Comment (programmer-ke) [reply rotkehlchen/tests/unit/test_inquirer.py:561]: I think the type checker doesn't check tests, so this wasn't flagged

File: rotkehlchen/utils/mixins/penalizable_oracle.py
Comment (yabirgb) [reply rotkehlchen/utils/mixins/penalizable_oracle.py:21]: As options you can:

- Create an overrade of the method so when the literal `oracle_penalty_threshold_count` is provided the return type is int.
- Use a `cast()`
- Use a `# type: ignore` explaining why

I would do the first or the last (probably the last and 1 in a smaller refactor) because the cast has a small overhead in production

File: rotkehlchen/tests/unit/test_inquirer.py
Comment (yabirgb) [reply rotkehlchen/tests/unit/test_inquirer.py:561]: Yes, for non typed environments mypy doens't check on the tests

**Comment by programmer-ke** [comment]
> Also @programmer-ke try to squash your commits. We keep 1 commit per change. So no smaller commits updating things that were added in the same PR

Cool, I can squash the commits before the finally merging to the base branch

File: rotkehlchen/tests/api/test_settings.py
Comment (programmer-ke) [reply rotkehlchen/tests/api/test_settings.py]: I took this approach, but ended up having to create a new fixture to control the relevant flag

**Comment by programmer-ke** [comment]
ping @yabirgb, just a reminder

**Comment by programmer-ke** [comment]
I see an api test failed. From the message, it looks like it could be a throttling issue on Coingecko

**Comment by yabirgb** [comment]
> I see an api test failed. From the message, it looks like it could be a throttling issue on Coingecko

Yes don't worry! I've pinged @LefterisJP to do a final review since I don't have any more comments

File: rotkehlchen/db/settings.py
Comment (LefterisJP) [inline rotkehlchen/db/settings.py:411]: Undo this change. This should only get ModifiableDBSettings.

File: rotkehlchen/rotkehlchen.py
Comment (LefterisJP) [inline rotkehlchen/rotkehlchen.py:336]: oooof so it was never initialized? Damn ...

So the main thing here is the logic needs changing.

`update_entries` should only get a `ModifiableDBSettings` argument. So here you need to turn the returned settings to a ModifiableDBSettings or just more simply create an `initialize()` function which just does what `reset()` does but with an argument of `DBSettings`.

**Review by LefterisJP — COMMENTED** [review]
Some comments

**Comment by programmer-ke** [comment]
Hey @LefterisJP , I've pushed an update addressing your comments.

File: docs/api.rst
Comment (LefterisJP) [inline docs/api.rst:614]: still not aligned

**Review by LefterisJP — COMMENTED** [review]
All looks good. The small tiny alignment thing I will push myself since it seems CI needs approval again.

**Review by LefterisJP — APPROVED** [review]
LGTM. Failed test is due to bitmex sandbox changing their data and is being fixed in another PR.

File: docs/api.rst
Comment (programmer-ke) [reply docs/api.rst]: lol, I applied you patch. I thought it'd be fixed

**Comment by programmer-ke** [comment]
Thanks for the POAP, first time I'm getting this. Interesting concept

**Pull request merged**
